File: packages/agent-core/src/learning/graph.ts
/**
 * In-memory memory graph builder.
 *
 * Derives nodes from the bounded memory store and session store, and edges
 * from keyword overlap. This is a lightweight, in-process approximation of the
 * Python `learning_graph.py` algorithm used by `/journey` and `/memory-graph`.
 */

import type { BoundedMemoryStore } from "../memory/store.js";
import type { MemoryScope } from "../memory/types.js";
import type { AgentSession, SessionStore } from "../session/store.js";
import type { MemoryGraph, MemoryGraphEdge, MemoryGraphNode } from "./types.js";

export interface MemoryGraphBuilderOptions {
  /** Bounded memory store (read-only). */
  memoryStore: BoundedMemoryStore;
  /** Session store (read-only). */
  sessionStore: SessionStore;
  /** Maximum keyword edges emitted per source node. */
  maxEdgesPerNode?: number;
  /** Minimum keyword overlap score required to emit an edge. */
  minEdgeScore?: number;
}

function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\u4e00-\u9fff]+/g, " ")
    .split(/\s+/)
    .filter((t) => t.length > 2);
}

function sharedKeywordScore(a: string[], b: string[]): number {
  const setB = new Set(b);
  const shared = a.filter((t) => setB.has(t)).length;
  const total = new Set([...a, ...b]).size;
  return total === 0 ? 0 : shared / total;
}

interface ScoredEdge {
  target: string;
  score: number;
}

function topEdges(
  sourceId: string,
  candidates: ScoredEdge[],
  kind: MemoryGraphEdge["kind"],
  maxEdges: number,
): MemoryGraphEdge[] {
  const sorted = candidates.filter((c) => c.score > 0).sort((a, b) => b.score - a.score);
  return sorted.slice(0, maxEdges).map((c) => ({
    source: sourceId,
    target: c.target,
    kind,
    weight: Number(c.score.toFixed(4)),
  }));
}

/**
 * Build an in-memory memory graph from memory entries and sessions.
 */
export async function buildMemoryGraph(opts: MemoryGraphBuilderOptions): Promise<MemoryGraph> {
  const maxEdges = Math.max(0, opts.maxEdgesPerNode ?? 4);
  const minScore = opts.minEdgeScore ?? 0;
  const nodes: MemoryGraphNode[] = [];
  const edges: MemoryGraphEdge[] = [];

  const memoryNodes: MemoryGraphNode[] = [];
  const memoryTokens: string[][] = [];

  for (const scope of ["memory", "user"] as MemoryScope[]) {
    const entries = opts.memoryStore.getEntries(scope);
    for (let i = 0; i < entries.length; i++) {
      const entry = entries[i];
      if (!entry) continue;
      const id = `memory:${scope}:${i}`;
      const label = entry.content.slice(0, 60).replace(/\s+/g, " ").trim();
      memoryNodes.push({
        id,
        kind: "memory",
        label: label || id,
        source: scope,
        metadata: { importance: entry.importance, fullLength: entry.content.length },
      });
      memoryTokens.push(tokenize(entry.content));
    }
  }
  nodes.push(...memoryNodes);

  let sessions: AgentSession[] = [];
  try {
    sessions = await opts.sessionStore.listSessions();
  } catch {
    sessions = [];
  }

  for (const session of sessions) {
    nodes.push({
      id: `session:${session.id}`,
      kind: "session",
      label: session.title,
      timestamp: session.startedAt,
      source: session.source ?? "unknown",
      metadata: {
        messageCount: session.messageCount,
        toolCallCount: session.toolCallCount,
      },
    });
  }

  // Keyword edges among memory nodes.
  for (let i = 0; i < memoryNodes.length; i++) {
    const sourceNode = memoryNodes[i];
    if (!sourceNode) continue;
    const candidates: ScoredEdge[] = [];
    for (let j = 0; j < memoryNodes.length; j++) {
      if (i === j) continue;
      const targetNode = memoryNodes[j];
      if (!targetNode) continue;
      const score = sharedKeywordScore(memoryTokens[i] ?? [], memoryTokens[j] ?? []);
      if (score >= minScore) {
        candidates.push({ target: targetNode.id, score });
      }
    }
    edges.push(...topEdges(sourceNode.id, candidates, "keyword", maxEdges));
  }

  // Session -> memory edges by keyword overlap.
  for (const session of sessions) {
    const sessionId = `session:${session.id}`;
    const sessionText = `${session.title} ${session.preview ?? ""}`;
    const sessionTokens = tokenize(sessionText);
    const candidates: ScoredEdge[] = [];
    for (let i = 0; i < memoryNodes.length; i++) {
      const memoryNode = memoryNodes[i];
      if (!memoryNode) continue;
      const score = sharedKeywordScore(sessionTokens, memoryTokens[i] ?? []);
      if (score >= minScore) {
        candidates.push({ target: memoryNode.id, score });
      }
    }
    edges.push(...topEdges(sessionId, candidates, "related", maxEdges));
  }

  const stats = {
    nodeCount: nodes.length,
    edgeCount: edges.length,
    topicCount: 0,
    memoryCount: memoryNodes.length,
    sessionCount: sessions.length,
  };

  return { nodes, edges, stats };
}

File: src/schema/graph.rs
//! Serde types for the memory-graph builder IPC boundary.
//!
//! Mirrors `packages/agent-core/src/learning/graph.ts` + `learning/types.ts`
//! (MemoryGraph, MemoryGraphNode, MemoryGraphEdge, MemoryGraph stats).

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MemoryGraphRequest {
    #[serde(default)]
    pub memory_entries: Vec<MemoryEntry>,
    #[serde(default)]
    pub sessions: Vec<SessionEntry>,
    #[serde(default)]
    pub max_edges_per_node: Option<usize>,
    #[serde(default)]
    pub min_edge_score: Option<f64>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MemoryEntry {
    #[serde(default)]
    pub id: String,
    pub content: String,
    #[serde(default)]
    pub importance: f64,
    #[serde(default)]
    pub scope: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SessionEntry {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub preview: Option<String>,
    #[serde(default)]
    pub source: Option<String>,
    pub started_at: i64,
    pub message_count: usize,
    pub tool_call_count: usize,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MemoryGraph {
    pub nodes: Vec<MemoryGraphNode>,
    pub edges: Vec<MemoryGraphEdge>,
    pub stats: MemoryGraphStats,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MemoryGraphNode {
    pub id: String,
    pub kind: String,
    pub label: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MemoryGraphEdge {
    pub source: String,
    pub target: String,
    pub kind: String,
    pub weight: f64,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MemoryGraphStats {
    pub node_count: usize,
    pub edge_count: usize,
    pub topic_count: usize,
    pub memory_count: usize,
    pub session_count: usize,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn request_round_trips() {
        let req: MemoryGraphRequest = serde_json::from_value(serde_json::json!({
            "memoryEntries": [{ "content": "hello", "scope": "memory" }],
            "sessions": []
        }))
        .unwrap();
        assert_eq!(req.memory_entries.len(), 1);
        assert_eq!(req.memory_entries[0].scope, "memory");
        assert!(req.max_edges_per_node.is_none());
    }
}
